ArgsParsing/ArgsParseResult.cs

```diff
@@ -0,0 +1,166 @@
+        public static ArgsParseResult<T> Success(
+            Failure? nestedFailure,
+            T result,
+            IImmutableList<string> remainingArgs)
+        {
+            return new ArgsParseResult<T>(true, new Success<T>(result, remainingArgs), nestedFailure);
+        }
+
+        /// <summary>
+        /// Create a successful parse result object, without any nested failure.
+        /// </summary>
+        /// <param name="result">The successfully parsed object.</param>
+        /// <param name="remainingArgs">The remaining arguments that were not consumed.</param>
+        /// <returns>An respective instance of <see cref="ArgsParseResult{T}"/></returns>
+        public static ArgsParseResult<T> Success(
+            T result,
+            IImmutableList<string> remainingArgs)
+        {
+            return new ArgsParseResult<T>(true, new Success<T>(result, remainingArgs), null);
+        }
+
+        /// <summary>
+        /// Create an unsuccessful parse result object, without any nested failure.
+        /// </summary>
+        /// <param name="message">A message describing the error.</param>
+        /// <param name="relevance">(optional) How likely it is that this error is relevant to the overall parsing,
+        /// or in other words how useful it will probably be if reported to the user.</param>
+        /// <returns>An respective instance of <see cref="ArgsParseResult{T}"/></returns>
+        public static ArgsParseResult<T> Failure(
+            string message,
+            ErrorRelevanceConfidence relevance = ErrorRelevanceConfidence.Default)
+        {
+            return new ArgsParseResult<T>(false, default!, new Failure(relevance, message));
+        }
+
+        /// <summary>
+        /// Create an unsuccessful parse result object from an existing failure.
+        /// </summary>
+        /// <param name="failure">The existing failure to re-use the failure from.</param>
+        /// <returns>An respective instance of <see cref="ArgsParseResult{T}"/></returns>
+        public static ArgsParseResult<T> Failure(Failure? failure)
+        {
+            return new ArgsParseResult<T>(false, default!, failure);
+        }
+
+        /// <summary>
+        /// Accesses the parsing result, if <see cref="IsSuccess"/> is true.
+        /// </summary>
+        /// <param name="result">The contained parsing result.</param>
+        /// <param name="remainingArgs">The remaining, unconsumed arguments.</param>
+        /// <returns>Whether the result is successful and the out variable therefore set.</returns>
+        public bool TryUnwrap(out T result, out IImmutableList<string> remainingArgs)
+        {
+            if (IsSuccess)
+            {
+                result = SuccessResult.Result;
+                remainingArgs = SuccessResult.RemainingArgs;
+                return true;
+            }
+            else
+            {
+                result = default!;
```

Comment:
`default` _is_ `null` for every reference type... (this is the kind of silliness that non-nullable types create)

Reply:
yes but the field is not marked as null, and we have nullable reference types enabled, so assigning a null value to a non-nullable type should be very much avoided

Reply:
here the caller must check for the boolean result of the `TryUnwrap`. I don't want callers to check for null if the `TryUnwrap` succeeded, since the result will always be nonnull. I figure if a caller accesses the out variable after the boolean result was false, the program may rightfully blow up into his face

Reply:
The C-like approach of "do nonsense at your own risk" isn't compatible at all with static checking...

Reply:
of course it isn't, but I don't know how C# would let you do this safely without requiring a redundant null check

Reply:
i could have a method take an action `IfSuccess(Action<SuccessResult>)` that only gets executed if the result was successful, but that usually makes branching and getting stuff out of the action annoying to deal with

Reply:
is this resolved?

Reply:
found this on StackOverflow:

>Lucas Trzesniewski: The bool TrySomething(out Result result) pattern predates nullable structs in the langugage, and IMHO shouldn't be used for new code as it makes it inconventient for the caller (because of the need to declare an additional variable most of the time).
Using a nullable struct as a return value is much nicer to the caller, and doesn't require you to return meaningless values in your code.

I'll have a look at changing the api